### ISSUE TITLE
Use deep copy for setting internal slots?

### ISSUE THREAD

**Issue by zolkis** [webmachinelearning/webnn#395]
Should we do deep copy of the descriptor? Would the change of original operand's descriptor impact the copied one?

_Originally posted by @huningxin in https://github.com/webmachinelearning/webnn/pull/348#discussion_r1213183583_

**Comment by zolkis** [comment]
[ECMAScript](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots) doesn't really shed light on this. Internal slots are supposed to be created and then not modified unless specified differently.

Now we seem to have 2 options:
1. change every line that sets an internal slot to replace `the argument` with "a copy of `the argument`"
2. or, create steps to "set the internal slot `[[name]]` to `<arg>`" steps, where we say "set `[[name]]` to a copy of `<arg>`".

@anssiko @dontcallmedom  any input here?

**Comment by anssiko** [comment]
If we need to do this in many places maybe it makes sense to define an abstract operation for doing a deep copy? Did you check whether any of the infra specs incl. HTML have reusable helpers for this?

For some inspiration, for structured data there's a bunch of abstract operations in HTML https://html.spec.whatwg.org/multipage/structured-data.html referenced by other specs such as IndexedDB, Web Workers.

**Comment by zolkis** [comment]
Yes, we can define separate "set internal slot" steps for that and when it becomes specified in INFRA or ECMA, we can switch.

What we need is to prevent changes from client after setting, i.e. to pass ownership of the init dict to the internal slot's object. An emulation of that is [deep] copy / pass by value, usually referenced as "copy of" in WebIDL, as suggested above.
Is that (option 2) good to go?

**Comment by anssiko** [comment]
Yes, I think a PR for that option would help.

**Comment by zolkis** [comment]
Since it affects all internal slots and the whole spec, I will do it after the current batch of PRs is merged.
Maybe after the stylistic changes.

**Comment by zolkis** [comment]
Setting internal slots has been explained [here](https://github.com/webmachinelearning/webnn/issues/450#issuecomment-1683052506).

**Comment by inexorabletash** [comment]
I've been going through the spec and I don't see anywhere where an internal slot holds a non-primitive data type, and the properties are mutated after it is set. Most specifically, once a MLOperand's [[descriptor]] is set I never see it changing.

The closest I've spotted is in [resample output sizes](https://webmachinelearning.github.io/webnn/#mlgraphbuilder-resample-output-sizes) where the input descriptor is cloned then mutated with:

> Let desc be an MLOperandDescriptor initialized to input.[[descriptor]].

This could be made more explicit with "Let desc be a new MLOperandDescriptor..."

Otherwise... I haven't spotted any usage that would require more explicit deep copies. Am I missing any?

FWIW, I was "grepping the spec" using this on the console:
```
console.log([...document.body.innerText.matchAll(/set .+?\[\[\w+\]\]\..+ to .*/ig)].map(m => m[0]).join('\n'))
```

**Comment by inexorabletash** [comment]
The [calculate matmul output sizes](https://webmachinelearning.github.io/webnn/#mlgraphbuilder-calculate-matmul-output-sizes) may also be misleading, as it does:

> Let shapeA be a.[[descriptor]].dimensions  ...
> ... then insert 1 in the front of shapeA ...

The intent here is a copy, not mutating the input's descriptor.

**Comment by inexorabletash** [comment]
@zolkis - do you think we can close this out? Or can you spot any other changes we should make?

**Comment by zolkis** [comment]
@inexorabletash yes, we can close it, thanks for the lots of help here.
